README.md
ImageToHacklace2
================

This tool converts images files into a string of hexadecimal bytes which can be downloaded as part of an animation onto the "HackLace 2" (A Necklace for Hackers - see [1]).

**supported image formats**

- Bitmap (BMP)
- Graphics Interchange Format (GIF)
- Portable Network Graphics (PNG)
- JPEG File Interchange Format (JPG / JPEG) - not recommended, based on the compression rate and hence artefacts the result might not look as you expect

A pixel with white color is converted as "LED = on" any other pixel is converted as "LED = off". Best results you get with monochrom images.

**supported image size**

- width: 1 <= x <= 255
- height: 1 <= y <= 8

If the image is greater in one dimension then this dimension will be truncated to the maximum value. The origin of ordinates is the upper left corner of the image. The height will be converted always as 8 pixel height. If the height of the input image is less than 8 pixel, the missed pixel converted as "LED = off".

**usage - convert a single image**

```
java -jar ImageToHacklace2.jar image_file
```

The output, for example might look like `\1F 08 00 20 42 40 5C 40 42 20\`, you can use as datablock for the Hacklace 2 animation app.

see an example for the Linux shell:

```
printf '\eHL\\00 70 01 1F 08 00 20 42 40 5C 40 42 20\\ \\00 1F FF\\' > /dev/ttyUSB0
```

For further information about the above command have a look on the Hacklace 2 wiki[1].

**usage - convert a template file**

A template file is a Hacklace command script or a Shell script inlcuding a output command (see above example) which includes a placeholders for an image file. If the image file from the placeholderncan be found the placeholder is substituted by hexadecimal bytes representing the image file.

<u>template_example.txt</u>
```
HL
\00 7A 01\ It's hip to be [[square.png]] ! \00\
\00 7A 01\ It's hip to [[smile.png]] ! \00\
\FF\
```

```
java -jar ImageToHacklace2.jar --template template_example.txt
```

Assuming the image file `square.png` exist in the current directory and the image file `smile.png` does not exist. The file `template_example.txt` will be converted to `template_example.hl` like below.

<u>template_example.hl</u>
```
HL
\00 7A 01\ It's hip to be \1F 08 00 7E 42 42 42 42 7E 00\ ! \00\
\00 7A 01\ It's hip to [[smile.png]] ! \00\
\FF\
```

For further information about the format of a hacklace script have a look on the Hacklace 2 wiki[1].


[1] http://wiki.fab4u.de/wiki/Hacklace


changelog.md
**version 0.1.1**
- add processing of Hacklace template files
- documentation updated

**version 0.1.0**
- initial version

src/test/java/sub/optimal/hacklace2/ConverterTest.java
/*
 * Copyright (C) 2014 suboptimal
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package sub.optimal.hacklace2;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.IOException;
import java.io.OutputStreamWriter;
import org.junit.Assert;
import junit.framework.TestSuite;
import org.junit.Test;

/**
 *
 * @author suboptimal
 */
public class ConverterTest extends TestSuite {

    Converter converter = Converter.getInstance();

    @Test
    public void testDotUpperLeft() {
        String fileName = "dot_upper_left.gif";
        String expectedHl = "\\1F 08 01 00 00 00 00 00 00 00\\";
        runConvertToHacklaceTest(fileName, expectedHl);
    }

    @Test
    public void testDotLowerLeft() {
        String fileName = "dot_lower_left.gif";
        String expectedHl = "\\1F 08 80 00 00 00 00 00 00 00\\";
        runConvertToHacklaceTest(fileName, expectedHl);
    }

    @Test
    public void testDotLowerRight() {
        String fileName = "dot_lower_right.gif";
        String expectedHl = "\\1F 08 00 00 00 00 00 00 00 80\\";
        runConvertToHacklaceTest(fileName, expectedHl);
    }

    @Test
    public void testDotUpperRight() {
        String fileName = "dot_upper_right.gif";
        String expectedHl = "\\1F 08 00 00 00 00 00 00 00 01\\";
        runConvertToHacklaceTest(fileName, expectedHl);
    }

    @Test
    public void testPattern5A() {
        String fileName = "pattern_5A.gif";
        String expectedHl = "\\1F 08 5A 5A 5A 5A 5A 5A 5A 5A\\";
        runConvertToHacklaceTest(fileName, expectedHl);
    }

    @Test
    public void testPatternA5() {
        String fileName = "pattern_A5.gif";
        String expectedHl = "\\1F 08 A5 A5 A5 A5 A5 A5 A5 A5\\";
        runConvertToHacklaceTest(fileName, expectedHl);
    }

    @Test
    public void testPattern0F() {
        String fileName = "pattern_0F.gif";
        String expectedHl = "\\1F 08 0F 0F 0F 0F 0F 0F 0F 0F\\";
        runConvertToHacklaceTest(fileName, expectedHl);
    }

    @Test
    public void testPatternF0() {
        String fileName = "pattern_F0.gif";
        String expectedHl = "\\1F 08 F0 F0 F0 F0 F0 F0 F0 F0\\";
        runConvertToHacklaceTest(fileName, expectedHl);
    }

    @Test
    public void testPatternFF() {
        String fileName = "pattern_FF.gif";
        String expectedHl = "\\1F 08 FF FF FF FF FF FF FF FF\\";
        runConvertToHacklaceTest(fileName, expectedHl);
    }

    @Test
    public void testOutOfRangeHeight() {
        String fileName = "out_of_range_height.gif";
        String expectedHl = "\\1F 08 01 02 04 08 10 20 40 80\\";
        runConvertToHacklaceTest(fileName, expectedHl);
    }

    @Test
    public void testOutOfRangeHeight4() {
        String fileName = "out_of_range_height_4.gif";
        String expectedHl = "\\1F 08 08 04 02 01 01 02 04 08\\";
        runConvertToHacklaceTest(fileName, expectedHl);
    }

    @Test
    public void testOutOfRangeWidth() {
        String fileName = "out_of_range_width.gif";
        StringBuilder expectedHl = new StringBuilder("\\1F FF");
        for (int i = 0; i < 31; i++) {
            expectedHl.append(" 01 02 04 08 10 20 40 80");
        }
        expectedHl.append(" 01 02 04 08 10 20 40");
        expectedHl.append("\\");
        runConvertToHacklaceTest(fileName, expectedHl.toString());
    }

    @Test
    public void testTemplateProcessorExistingImageFile() {
        try {
            String templateBasename = "template_existing_image_file";
            File hacklaceFile = runTemplateProcessorTest(templateBasename);
            assertTemplateProcessortTest(templateBasename, hacklaceFile);
        } catch (IOException ex) {
            System.err.println("no exception expected: " + ex.getMessage());
            Assert.fail();
        }
    }

    @Test
    public void testTemplateProcessorNotAnImageFile() {
        try {
            String templateBasename = "template_not_an_image_file";
            File hacklaceFile = runTemplateProcessorTest(templateBasename);
            assertTemplateProcessortTest(templateBasename, hacklaceFile);
        } catch (IOException ex) {
            System.err.println("no exception expected: " + ex.getMessage());
            Assert.fail();
        }
    }

    @Test
    public void testTemplateProcessorNotExistingImageFile() {
        try {
            String templateBasename = "template_not_existing_image_file";
            File hacklaceFile = runTemplateProcessorTest(templateBasename);
            assertTemplateProcessortTest(templateBasename, hacklaceFile);
        } catch (IOException ex) {
            System.err.println("no exception expected: " + ex.getMessage());
            Assert.fail();
        }
    }

    /**
     * Compares the expected output file wth the actual created output file.
     *
     * @param templateBasename basename of the template file
     * @param hacklaceFile {@link File} object representing the Hacklace file which is created
     * during the conversion of the template
     * @throws IOException in case reading from one file is failed
     */
    private void assertTemplateProcessortTest(String templateBasename, File hacklaceFile) throws IOException {
        File expectedOutput = new File(getExpectedResultFilename(templateBasename));
        BufferedReader expectedHl = new BufferedReader(new FileReader(expectedOutput));
        BufferedReader actualHl = new BufferedReader(new FileReader(hacklaceFile));
        String inLine;
        String outLine;
        do {
            inLine = expectedHl.readLine();
            outLine = actualHl.readLine();
            Assert.assertEquals(inLine, outLine);

        } while (inLine != null && outLine != null);
    }

    /**
     * Executes the template processor test.
     *
     * @param templateBasename basename of the template file
     * @return a {@link File} object representing the Hacklace file which is created during the
     * conversion of the template
     * @throws IOException
     */
    private File runTemplateProcessorTest(String templateBasename) throws IOException {
        File testInput = new File(getTemplateName(templateBasename));
        File hacklaceFile = new File(getHacklaceName(templateBasename));
        if (hacklaceFile.exists()) {
            hacklaceFile.delete();
        }
        TemplateProcessor templateProcessor = TemplateProcessor.getInstance();
        templateProcessor.processTemplateFile(testInput);
        hacklaceFile.deleteOnExit();
        return hacklaceFile;
    }

    /**
     * Executes the convert to hacklace test.
     * @param imageFileName name of the image file to convert
     * @param expectedHl the expected Hacklace hexadecimal bytes string
     */
    private void runConvertToHacklaceTest(String imageFileName, String expectedHl) {
        File image = new File(imageFileName);
        String hlOutput = converter.convertToHacklace(image);
        Assert.assertEquals(expectedHl, hlOutput);
    }

    private String getTemplateName(String templateBasename) {
        return templateBasename + ".txt";
    }

    private String getHacklaceName(String templateBasename) {
        return templateBasename + ".hl";
    }

    private String getExpectedResultFilename(String templateBasename) {
        return templateBasename + ".expected";
    }
}


src/main/java/sub/optimal/util/FileUtil.java
/*
 * Copyright (C) 2014 suboptimal
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package sub.optimal.util;

import java.io.File;

/**
 *
 * @author suboptimal
 */
public class FileUtil {

    /**
     * Check if the passed {@link File} point to an existing file and can be read by the user.
     *
     * @param file file to be checked
     * @return true - <code>file</code> point to a file and is readable<br>
     * false - <code>file</file> is either a directory or is not readable
     */
    public static boolean isReadableFile(File file) {
        if (!file.exists() || !file.isFile()) {
            System.err.println(String.format("%s: does not exist", file.getName()));
            return false;
        }
        if (!file.canRead()) {
            System.err.println(String.format("%s: no read permission", file.getName()));
            return false;
        }
        return true;
    }
}


src/main/java/sub/optimal/hacklace2/Converter.java
/*
 * Copyright (C) 2014 suboptimal
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package sub.optimal.hacklace2;

import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import javax.imageio.ImageIO;

/**
 *
 * @author suboptimal
 */
public class Converter {

    private static final String HEX_BYTE_MARKER = "\\";
    public static final String DATABLOCK_START = HEX_BYTE_MARKER + "1F";
    public static final int RGB_WHITE = 16777215;
    private static final String ZERO_LENGTH_STRING = "";

    private Converter() {
    }

    /**
     * Return an instance of the converter.
     *
     * @return new instance of the {@link Converter}
     */
    public static Converter getInstance() {
        return new Converter();
    }

    /**
     * Return the marker which indicate the start/end of hexadecimal bytes in the Hacklace command.
     *
     * @return the hexadecimal bytes marker
     */
    public static String getHexByteMarker() {
        return HEX_BYTE_MARKER;
    }

    /**
     * Check if the passed image dimension are in the supported output range.
     *
     * @param height heigth of the image
     * @param width width of the image
     * @return false - if the dimension is below the minimal supported dimension<br>
     * true - if the dimension is in the supported range, in case the dimension is out the maximum
     * convertable range an warning is printed
     */
    private boolean checkDimension(int height, int width) {
        if (height < 1 || width < 1) {
            System.err.println("image dimension out of range: minimal dimension = 1 x 1 pixel");
            return false;
        }
        if (height > 8) {
            System.err.println("image height out of range: area will be truncated to 8 pixel height");
        }
        if (width > 255) {
            System.err.println("image width out of range: area will be truncated to 255 pixel width");
        }
        return true;
    }

    /**
     * Convert the passed image file into a string of hexadecimal bytes which can be passed to the
     * Hacklace 2 animation app.
     *
     * @param image image file
     * @return the hexadecimal bytes which represents the passed image
     */
    public String convertToHacklace(File image) {
        try {
            BufferedImage buffImage = ImageIO.read(image);
            if (buffImage == null) {
                System.err.println(image.getName() + ": not supported image file format");
                return ZERO_LENGTH_STRING;
            }
            int height = buffImage.getHeight();
            int width = buffImage.getWidth();
            if (!checkDimension(height, width)) {
                return ZERO_LENGTH_STRING;
            }
            width = (width < 255 ? width - 1 : 254);
            height = (height < 8 ? height - 1 : 7);
            StringBuilder out = new StringBuilder();
            out.append(String.format("%s %02X", DATABLOCK_START, width + 1));
            for (int column = 0; column <= width; column++) {
                int hlValue = 0;
                for (int row = height; row >= 0; row--) {
                    hlValue <<= 1;
                    if ((buffImage.getRGB(column, row) & RGB_WHITE) != RGB_WHITE) {
                        hlValue++;
                    }
                }
                out.append(String.format(" %02X", hlValue));
            }
            out.append(HEX_BYTE_MARKER);
            return out.toString();
        } catch (IOException ex) {
            System.err.println(image.getName() + ": failure during processing - " + ex.getMessage());
            return ZERO_LENGTH_STRING;
        }
    }
}


src/main/java/sub/optimal/hacklace2/TemplateProcessor.java
/*
 * Copyright (C) 2014 suboptimal
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package sub.optimal.hacklace2;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import sub.optimal.util.FileUtil;

/**
 *
 * @author suboptimal
 */
public class TemplateProcessor {

    /**
     * Pattern to match an image placeholder in the format <code>[[filename]]</code>.
     */
    private static final Pattern placeholder = Pattern.compile("\\[\\[([^\\]]*)\\]\\]");
    private Converter converter;

    private TemplateProcessor() {
    }

    /**
     * Return an instance of the converter.
     *
     * @return new instance of the {@link Converter}
     */
    public static TemplateProcessor getInstance() {
        return new TemplateProcessor();
    }

    /**
     * Process a Hacklace script template. Placeholder for existing image files will be substituted.
     *
     * @param template Hacklace template file
     * @throws IOException if reading from the template file or writing to the output file fails
     */
    public void processTemplateFile(File template) throws IOException {
        converter = Converter.getInstance();

        if (!FileUtil.isReadableFile(template)) {
            System.err.println(String.format("%s: cannot be accessed", template.getName()));
            return;
        }

        String templateFileName = template.getName();
        int indexLastDot = templateFileName.lastIndexOf(".");
        String baseName = templateFileName.substring(0, indexLastDot < 0 ? templateFileName.length() : indexLastDot);
        File hacklaceFile = new File(baseName + ".hl");
        if (hacklaceFile.exists()) {
            System.err.println(String.format("%s: output file already exist", hacklaceFile.getName()));
            return;
        }

        BufferedReader reader = null;
        BufferedWriter writer = null;
        try {
            reader = new BufferedReader(new InputStreamReader(new FileInputStream(template), "ISO-8859-1"));
            writer = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(hacklaceFile), "ISO-8859-1"));

            String s;
            while ((s = reader.readLine()) != null) {
                writer.write(replacePlaceholder(s));
                // TODO check for DOS/UNIX lineends
                writer.write('\n');
            }
        } finally {
            if (writer != null) {
                try {
                    writer.close();
                } catch (IOException ioe) {
                    System.err.println(String.format("%s: failed to close the file - %s", hacklaceFile.getName(), ioe.getMessage()));
                }
            }
            if (reader != null) {
                try {
                    reader.close();
                } catch (IOException ioe) {
                    System.err.println(String.format("%s: failed to close the file - %s", hacklaceFile.getName(), ioe.getMessage()));
                }
            }
        }
    }

    /**
     * Replace an image placeholder in the passed line by hexadecimal bytes in Hacklace 2 animation
     * format.<br>
     * The placeholder has following format: [[filename]]<br>
     * - if <code>filename</code> point to a readable file, the placeholder is substituted
     *
     * @param line input string
     * @return output line, placeholder for existing image files are substituted
     */
    public String replacePlaceholder(String line) {
        Matcher matcher = placeholder.matcher(line);
        StringBuffer lineBuffer = new StringBuffer();
        StringBuilder matchBuffer = new StringBuilder();
        while (matcher.find()) {
            matchBuffer.setLength(0);
            String fileName = matcher.group(1);
            File file = new File(fileName);
            if (FileUtil.isReadableFile(file)) {
                String hacklaceBytes = converter.convertToHacklace(file);
                if (hacklaceBytes.length() != 0) {
                    System.out.println(String.format("%s: image converted", file.getName()));
                    matchBuffer.append(Converter.getHexByteMarker());
                    matchBuffer.append(hacklaceBytes);
                    matchBuffer.append(Converter.getHexByteMarker());
                } else {
                    matchBuffer.append(matcher.group());
                }
            } else {
                matchBuffer.append(matcher.group());
            }
            matcher.appendReplacement(lineBuffer, matchBuffer.toString());
        }
        matcher.appendTail(lineBuffer);
        return lineBuffer.toString();
    }
}


src/main/java/sub/optimal/Img2Hl.java
/*
 * Copyright (C) 2014 suboptimal
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package sub.optimal;

import java.io.File;
import java.io.IOException;
import java.util.logging.Level;
import java.util.logging.Logger;
import sub.optimal.hacklace2.Converter;
import sub.optimal.hacklace2.TemplateProcessor;
import sub.optimal.util.FileUtil;

/**
 * Main class of ImageToHacklace2 converter.<br><br>
 * This tool converts images files into a string of hexadecimal bytes which can be downloaded<br>
 * as part of an animation onto the "HackLace2" (see <a
 * href="http://wiki.fab4u.de/wiki/Hacklace">http://wiki.fab4u.de/wiki/Hacklace</a>).
 *
 * @author SubOptimal
 * @version 0.1.0
 */
public class Img2Hl {

    private static final String VERSION = "0.1.1";

    private static String templateFileName;
    private static String imageFileName;

    public static void main(String[] args) {
        processParameters(args);
        if (imageFileName != null) {
            processImageFile();
        } else if (templateFileName != null) {
            processTemplateFile();
        }
    }

    /**
     * Process the template file.
     */
    private static void processTemplateFile() {
        File template = new File(templateFileName);
        if (FileUtil.isReadableFile(template)) {
            TemplateProcessor converter = TemplateProcessor.getInstance();
            try {
                converter.processTemplateFile(template);
            } catch (IOException ex) {
                System.err.println("convertion failed: " + ex.getMessage());
            }
        }
    }

    /**
     * Process an single image file.
     */
    private static void processImageFile() {
        File image = new File(imageFileName);
        if (FileUtil.isReadableFile(image)) {
            Converter converter = Converter.getInstance();
            System.out.println(converter.convertToHacklace(image));
        }
    }

    /**
     * Process the passed application parameters.
     * @param args
     */
    private static void processParameters(String[] args) {
        if (args.length == 0) {
            showUsage();
            System.exit(1);
        }

        for (int i = 0; i < args.length; i++) {
            String opt = args[i];
            if (opt.startsWith("--template")) {
                if ((i + 1) < args.length) {
                    templateFileName = args[++i];
                } else {
                    System.err.println("parameter missed for option --template");
                    System.exit(1);
                    processTemplateFile();
                }
            } else if (opt.startsWith("--")) {
                System.err.printf("unknown option passed '%s'%n", opt);
                System.exit(1);
            } else {
                imageFileName = opt;
            }
        }
    }

    private static void showUsage() {
        System.err.printf("ImageToHacklace2 " + VERSION + "%n%n");
        System.err.printf("usage: java -jar ImageToHacklace2 [ image_file | --template template_file ]%n%n");
        System.err.printf("%16s%s%n", "image_file -", "an image file to convert into the hacklace databytes");
        System.err.printf("%16s%s%n", "", "recommended formats: BMP, GIF, PNG");
        System.err.printf("%16s%s%n", "--template -", "template hacklace configuration file with placeholders");
        System.err.printf("%16s%s%n", "", "for image files");
    }
}
